Fix myq increasing number of network connections

ehendrix23 opened the pull request:
## Description:
Updated pymyq version to 1.2.0. The aiohttp client session is now setup within the pymyq library instead with tested socket options to prevent continuous increase in sockets on certain platforms.

**Related issue (if applicable):** fixes #9193

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [X] There is no commented out code in this PR.

If the code communicates with devices, web services, or third-party tools:
  - [X] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.

ehendrix23 committed:
Fix for network issues

Fix for network issues

ehendrix23 committed:
Further changes to network connection

ehendrix23 committed:
websession is created in pymyq

websession is created in pymyq instead.
Added call on stop event to close web session.

ehendrix23 committed:
Updated requirements file

awarecan commented:
Can you elaborate what changes in your session fixed the leak socket issue? `enable_cleanup_closed=True`?

If so, maybe we should change `aiohttp_client` to default turn on this option.

ehendrix23 commented:
So this is what I have found out so far. This is experience and testing with aiohttp for 2 components that were having issues on certain deployments of HassIO.

Harmony component (through library aioharmony) is using websockets from aiohttp. People were having regular disconnects and then re-connections failing. Sometimes also with errors like "Network unreachable" etc... 
In that one a solution was to set the connection to be socket.AF_INET and verify_ssl=False. This is based on some info I found for aiohttp with others running into similar issues: 
https://github.com/aio-libs/aiohttp/issues/2522
https://github.com/aio-libs/aiohttp/issues/3171

The whole conversation related to aioharmony is here: https://community.home-assistant.io/t/harmony-hub-problems-after-update/98460

Next issue with myq. Somewhat similar but they would just see the number of connections keep on increasing even although it's supposed to keep the connection open and using pooling (all managed through aiohttp). 
Hence I started off with same settings. Issue was still there only now happening every 30 seconds or so instead of every 10 seconds (every 10 seconds a request is sent).
I then set limit_per_host to 5 and added enable_cleanup_closed=True.
After that change is was reported back to overall it only had 1 connection open, with a maximum of 2 connections. 
That is still below the limit_per_host of 5. 

What I currently do now know however is what if it is only enable_cleanup_closed=True and family is not set. I know with the issue on websockets I had to set it to resolve that issue.

awarecan commented:
I believe ` enable_cleanup_closed=True` is the key, can you try to only set it?

It seems happened in particular website, which didn't handle the SSL connection close gracefully.

pvizeli commented:
You can create a new websession with same connector pool with the helper. I would prefere that one because it need less resource and home assistant is care that this session will be closed: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/aiohttp_client.py#L49

ehendrix23 commented:
> You can create a new websession with same connector pool with the helper. I would prefere that one because it need less resource and home assistant is care that this session will be closed: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/aiohttp_client.py#L49

It's not about creating new sessions. It's about aiohttp "leaking" connections. Same connection pool is being used but and idea is same connection is used all the time. Yet on certain platforms they were seeing a constant increase in number of connections.
Was using the HA aiohttp client before to create the web session.

ehendrix23 commented:
Just checking 

> Can you elaborate what changes in your session fixed the leak socket issue? `enable_cleanup_closed=True`?
 
> If so, maybe we should change `aiohttp_client` to default turn on this option.
They confirmed this indeed helps.

ehendrix23 commented:
Not sure what needs to be done further to get this approved. I see a number of new checks sitting there though.

ehendrix23 committed:
Added comment

awarecan commented:
I think it has been confirmed that `enable_cleanup_closed=True` works, can you change the pull request to change aiohttp_client add this option instead of maintenance your own session.

ehendrix23 committed:
Changed back to use aiohttp_client

ehendrix23 committed:
Cleanup closed sockets in aiohttp

Enable automatic cleanup of closed sockets in aiohttp client helper.

ehendrix23 commented:
> I think it has been confirmed that `enable_cleanup_closed=True` works, can you change the pull request to change aiohttp_client add this option instead of maintenance your own session.

Changed it back to use aiohttp helper client from HA and updated aiohttp helper client to include this option.

ehendrix23 committed:
Updated manifest & requirements

awarecan commented on homeassistant/components/myq/cover.py:
```diff
@@ -1,16 +1,22 @@
+Support for MyQ-Enabled Garage Doors.
 
+For more details about this platform, please refer to the documentation
+https://home-assistant.io/components/cover.myq/
```
We don't use link in the file header anymore. Please revert the changes in the comment block.

awarecan commented:
@pvizeli 

Per aiohttp document
> enable_cleanup_closed (bool) – some SSL servers do not properly complete SSL shutdown process, in that case asyncio leaks ssl connections. If this parameter is set to True, aiohttp additionally aborts underlining transport after 2 seconds. It is off by default.

However add this option will apply to all clients, would you mind take a look into it?

ehendrix23 committed:
Updated comment block

balloob commented:
I think that it's fine.
